### PR TITLE
feat(proofs): proof negotiation

### PR DESCRIPTION
**Pull request by Przytua**
Signed-off-by: Łukasz Przytuła <lprzytula@gmail.com>

Allow proof negotiation by responding to presentation proposal with new presentation request, and responding to presentation request with new presentation proposal

### CODE REVIEW

File: packages/core/src/modules/proofs/ProofsApi.ts [hunk 1]
```diff
@@ -270,6 +317,46 @@ export class ProofsApi<
...
+
+    const requestOptions: CreateRequestAsResponseOptions<PFs> = {
+      proofRecord,
+      proofFormats: options.proofFormats,
+      autoAcceptProof: options.autoAcceptProof,
+      comment: options.comment,
+    }
+    const { message } = await service.createRequestAsResponse(this.agentContext, requestOptions)
+
+    const outbound = createOutboundMessage(connection, message)
```
Comment (genaris) [inline packages/core/src/modules/proofs/ProofsApi.ts:354]: createOutboundMessage has been replaced by the class OutboundMessageContext

**Review by genaris — COMMENTED** [review]
This is very interesting. But I'm curious about the usage of this negotiation in a loop that involves more than a single message of each type. I mean: is it possible for a proof exchange to follow the path: propose-presentation -> request-presentation -> propose-presentation -> request-presentation -> presentation? If so, what happens when we do `agent.proofs.findRequestMessage` or `agent.proofs.findProposalMessage`? As it is supposed to be a single message, would it be updated to the latest one?

**Comment by Przytua** [comment]
> is it possible for a proof exchange to follow the path: propose-presentation -> request-presentation -> propose-presentation -> request-presentation -> presentation?

I assume (according to this diagram: https://github.com/hyperledger/aries-rfcs/blob/main/features/0037-present-proof/README.md#choreography-diagram) that it should be possible. With a small patch of AFJ and current implementation of acapy we were able to achieve `request-presentation -> propose-presentation -> request-presentation -> presentation` and it was properly updating former proof request message with the new one updated according to presentation proposal.

> As it is supposed to be a single message, would it be updated to the latest one?

Yes. I added some more tests to make sure it returns latest message.

File: packages/core/src/modules/proofs/ProofsApi.ts [hunk 2]
```diff
@@ -191,6 +196,47 @@ export class ProofsApi<
...
+   */
+  public async proposeProofAsResponse(options: ProposeProofAsResponseOptions<PFs>): Promise<ProofExchangeRecord> {
+    const { proofRecordId } = options
+    const proofRecord = await this.getById(proofRecordId)
+
+    const service = this.getService(proofRecord.protocolVersion)
+
+    if (!proofRecord.connectionId) {
+      throw new AriesFrameworkError(
+        `No connectionId found for credential record '${proofRecord.id}'. Connection-less issuance does not support presentation proposal or negotiation.`
```
[the remark this answers is not in the document]
Comment (Przytua) [reply packages/core/src/modules/proofs/ProofsApi.ts]: Fixed. I copied it from the `acceptProposal` from the same file, so fixed it there as well.

File: packages/core/src/modules/proofs/ProofsApi.ts [hunk 3]
```diff
@@ -270,6 +317,46 @@ export class ProofsApi<
...
+  public async requestProofAsResponse(options: RequestProofAsResponseOptions<PFs>): Promise<ProofExchangeRecord> {
+    const { proofRecordId } = options
+
+    const proofRecord = await this.getById(proofRecordId)
+
+    const service = this.getService(proofRecord.protocolVersion)
+
+    if (!proofRecord.connectionId) {
+      throw new AriesFrameworkError(
+        `No connectionId found for credential record '${proofRecord.id}'. Connection-less issuance does not support negotiation.`
```
[the remark this answers is not in the document]
Comment (Przytua) [reply packages/core/src/modules/proofs/ProofsApi.ts]: Fixed

File: packages/core/src/modules/proofs/ProofsApi.ts (hunk 1 above)
Comment (Przytua) [reply packages/core/src/modules/proofs/ProofsApi.ts]: Fixed

File: packages/core/src/modules/proofs/ProofsApi.ts [hunk 4]
```diff
@@ -56,10 +59,12 @@ import { ProofRepository } from './repository/ProofRepository'
 export interface ProofsApi<PFs extends ProofFormat[], PSs extends ProofService<PFs>[]> {
   // Proposal methods
   proposeProof(options: ProposeProofOptions<PFs, PSs>): Promise<ProofExchangeRecord>
+  proposeProofAsResponse(options: ProposeProofAsResponseOptions<PFs>): Promise<ProofExchangeRecord>
```
[the remark this answers is not in the document]
Comment (Przytua) [reply packages/core/src/modules/proofs/ProofsApi.ts]: Renamed to `negotiateRequest` and `negotiateProposal` accordingly in the `ProofsApi`. Functions in services were already implemented, and I just used them in the api. They're named `createProposalAsResponse` and `createRequestAsResponse`. Do you want me to rename them as well or is that fine?

File: packages/core/src/modules/proofs/protocol/v2/__tests__/indy-proof-negotiation.test.ts [hunk 5]
```diff
@@ -0,0 +1,444 @@
+import type { Agent } from '../../../../../agent/Agent'
```
Comment (TimoGlastra) [inline packages/core/src/modules/proofs/protocol/v2/__tests__/indy-proof-negotiation.test.ts:1]: It seems there's thests both in the top level tests and also in the protocol specific directories. What's the difference between them?

File: packages/core/src/modules/proofs/ProofsApi.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/core/src/modules/proofs/ProofsApi.ts]: We probably want to rename them, but it's more for a different pr I think.

**Comment by TimoGlastra** [comment]
sorry forgot about this one. Code LGTM, just one qeustion about the tests

File: packages/core/src/modules/proofs/protocol/v2/__tests__/indy-proof-negotiation.test.ts (hunk 5 above)
Comment (Przytua) [reply packages/core/src/modules/proofs/protocol/v2/__tests__/indy-proof-negotiation.test.ts:1]: Yes, I added tests in the `core/tests` first (as it seemed to me more of a e2e test), and later I noticed, that there are test cases in the `__tests__` folder of the protocol, that are covering similar scenarios, so added there as well. But true, they're almost no different

Comment (TimoGlastra) [reply packages/core/src/modules/proofs/protocol/v2/__tests__/indy-proof-negotiation.test.ts:1]: Can we remove one of them in that case to keep this tidy? I think my preference would go to keep them in the credentials module folder

Comment (Przytua) [reply packages/core/src/modules/proofs/protocol/v2/__tests__/indy-proof-negotiation.test.ts:1]: I assume you mean proofs module folder. Anyway removed those in `core/tests`.

Comment (genaris) [reply packages/core/src/modules/proofs/protocol/v2/__tests__/indy-proof-negotiation.test.ts:1]: Yeah I think this comes from the proofs module, where its root __tests__ directory has some unit tests but protocol/vx includes e2e tests.

I agree on merge these tests and leaving one of each in modules/proofs/__tests__. They can be named 'xxx.e2e.test.ts' like it happens in other modules E2E tests. Later we should open a new PR to harmonize this structure.

Comment (TimoGlastra) [reply packages/core/src/modules/proofs/protocol/v2/__tests__/indy-proof-negotiation.test.ts:1]: > I assume you mean proofs module folder

yes 😬

**Comment by Przytua** [comment]
> @Przytua if you can update the branch with main I can merge

@TimoGlastra updated

**Pull request merged**